hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleContextImpl.java

```diff
@@ -599,18 +627,99 @@ private void dispatchSyntheticTxn(
                 feeManager,
                 exchangeRateManager,
                 userTransactionConsensusTime,
-                authorizer);
+                authorizer,
+                solvencyPreCheck);
 
         try {
             dispatcher.dispatchHandle(childContext);
             childRecordBuilder.status(ResponseCodeEnum.SUCCESS);
             childStack.commitFullStack();
-        } catch (HandleException e) {
+        } catch (final HandleException e) {
             childRecordBuilder.status(e.getStatus());
             recordListBuilder.revertChildrenOf(childRecordBuilder);
         }
     }
 
+    private void validate(
+            @NonNull final KeyVerifier verifier,
+            final HederaFunctionality function,
+            final TransactionBody txBody,
+            final AccountID payer,
+            final Key payerKey,
+            final TransactionCategory txCategory,
+            final long nodeID)
+            throws PreCheckException {
+
+        final PreHandleContextImpl preHandleContext;
+
+        preHandleContext = new PreHandleContextImpl(readableStoreFactory(), txBody, payer, configuration(), dispatcher);
+        dispatcher.dispatchPreHandle(preHandleContext);
+
+        // Check for duplicate transactions. It is perfectly normal for there to be duplicates -- it is valid for
+        // a user to intentionally submit duplicates to multiple nodes as a hedge against dishonest nodes, or for
+        // other reasons. If we find a duplicate, we *will not* execute the transaction, we will simply charge
+        // the payer (whether the payer from the transaction or the node in the event of a due diligence failure)
+        // and create an appropriate record to save in state and send to the record stream.
+        final var duplicateCheckResult = recordCache.hasDuplicate(txBody.transactionID(), nodeID);
+        if (duplicateCheckResult != NO_DUPLICATE && duplicateCheckResult != SAME_NODE) {
+            throw new PreCheckException(DUPLICATE_TRANSACTION);
+        }
+
+        // Check the status and solvency of the payer
+        try {
+
+            final var fee = dispatchComputeFees(txBody, payer);
+            final var payerAccount = solvencyPreCheck.getPayerAccount(readableStoreFactory(), payer);
+            solvencyPreCheck.checkSolvency(txBody, payer, functionality, payerAccount, fee, true);
+        } catch (final InsufficientServiceFeeException | InsufficientNonFeeDebitsException e) {
```

Comment:
`InsufficientServiceFeeException` and `InsufficientNonFeeDebitsException` are `PreCheckException`s. I don't think we need to catch them here.

Reply:
Noted